actors/evm/src/interpreter/precompiles/bls_util.rs

```diff
@@ -0,0 +1,422 @@
+
+    // Encode y coordinate
+    // Real part (y.fp[0])
+    fp_to_bytes(&mut out[2 * PADDED_FP_LENGTH..3 * PADDED_FP_LENGTH], &input.y.fp[0]);
+    // Imaginary part (y.fp[1])
+    fp_to_bytes(&mut out[3 * PADDED_FP_LENGTH..4 * PADDED_FP_LENGTH], &input.y.fp[1]);
+
+    out
+}
+
+/// Extracts a G2 point in Affine format from a 256 byte slice representation.
+///
+/// **Note**: This function will perform a G2 subgroup check if `subgroup_check` is set to `true`.
+///
+/// Subgroup checks are required for:
+/// - Scalar multiplication
+/// - Multi-scalar multiplication (MSM)
+/// - Pairing operations
+///
+/// But not required for:
+/// - Point addition
+/// - Point negation
+pub(super) fn extract_g2_input(
+    input: &[u8],
+    subgroup_check: bool,
+) -> Result<blst_p2_affine, PrecompileError> {
+    // Check input length (256 bytes = 4 * 64 bytes for x.re, x.im, y.re, y.im)
+    if input.len() != G2_INPUT_ITEM_LENGTH {
+        return Err(PrecompileError::IncorrectInputSize);
+    }
+
+    let [x_re, x_im, y_re, y_im] = remove_g2_padding(input)?;
+    // Convert bytes to point
+    let point = decode_g2_on_curve(x_re, x_im, y_re, y_im)?;
+
+    if subgroup_check {
+        // Subgroup check (more expensive but required for certain operations)
+        // Verifies that the point has the correct order and is in G2
+        // SAFETY: point is properly initialized above
+        unsafe {
+            if !blst_p2_affine_in_g2(&point) {
+                return Err(PrecompileError::InvalidInput);
+            }
+        }
+    } else {
+        // Basic curve check (less expensive, sufficient for addition)
+        // Only verifies that the point is on the curve
+        // SAFETY: point is properly initialized above
+        if unsafe { !blst_p2_affine_on_curve(&point) } {
+            return Err(PrecompileError::InvalidInput);
+        }
+    }
+
+    Ok(point)
+}
+
+/// https://eips.ethereum.org/EIPS/eip-2537
+/// Encodes a single finite field element into byte slice with padding.
+pub(super) fn fp_to_bytes(out: &mut [u8], input: *const blst_fp) {
+    if out.len() != PADDED_FP_LENGTH {
+        return;
+    }
+    let (padding, rest) = out.split_at_mut(PADDING_LENGTH);
+    padding.fill(0);
+    // SAFETY: Out length is checked previously, `input` is a blst value.
+    unsafe { blst_bendian_from_fp(rest.as_mut_ptr(), input) };
+}
+
+/// Extracts a G1 point in Affine format from a 128 byte slice representation.
+pub fn extract_g1_input(
+    input: &[u8],
+    subgroup_check: bool,
+) -> Result<blst_p1_affine, PrecompileError> {
+    if input.len() != G1_INPUT_LENGTH {
+        return Err(PrecompileError::IncorrectInputSize);
+    }
+
+    let [x_bytes, y_bytes] = remove_g1_padding(input)?;
+
+    let point = decode_g1_on_curve(x_bytes, y_bytes)?;
+
+    // Check if point is on curve (no subgroup check needed for addition)
+    if subgroup_check && unsafe { !blst_p1_affine_in_g1(&point) } {
```

Comment:
This comment does not match the code. We are in fact doing a subgroup check.

Reply:
Also this comment is referring to addition but this is the general utils package and these functions should be usable in any context.